rasa/engine/caching.py

```diff
@@ -0,0 +1,373 @@
+            )
+
+    def get_cached_output_fingerprint(self, fingerprint_key: Text) -> Optional[Text]:
+        """Retrieves fingerprint of output based on fingerprint key.
+
+        Args:
+            fingerprint_key: The fingerprint serves as key for the lookup of output
+                fingerprints.
+
+        Returns:
+            The fingerprint of a matching output or `None` in case no cache entry was
+            found for the given fingerprint key.
+        """
+        with self._sessionmaker.begin() as session:
+            query = sa.select(self.CacheEntry).filter_by(
+                fingerprint_key=fingerprint_key
+            )
+            match = session.execute(query).scalars().first()
+
+            if match:
+                # This result was used during a fingerprint run.
+                match.last_used = datetime.utcnow()
+                return match.output_fingerprint_key
+
+            return None
+
+    def get_cached_result(
+        self, output_fingerprint_key: Text
+    ) -> Tuple[Optional[Path], Optional[Type[Cacheable]]]:
+        """Returns a potentially cached output result.
+
+        Args:
+            output_fingerprint_key:
+
+        Returns:
+            - Path to directory containing the cached content or `None` in case no
+            persisted data was found
+            - Class of persisted data or None in case no persisted data was found.
+        """
+        result_location, result_type = self._get_cached_result(output_fingerprint_key)
+
+        if not result_location:
+            logger.debug(f"No cached output found for '{output_fingerprint_key}'")
+            return None, None
+
+        path_to_cached = Path(result_location)
+        if not path_to_cached.is_dir():
+            logger.debug(
+                f"Cached output for '{output_fingerprint_key}' can't be found on disk."
+            )
+            return None, None
+
+        return self._load_from_cache(result_location, result_type)
+
+    def _get_cached_result(
+        self, output_fingerprint_key: Text
+    ) -> Tuple[Optional[Path], Optional[Text]]:
+        with self._sessionmaker.begin() as session:
+            query = sa.select(
+                self.CacheEntry.result_location, self.CacheEntry.result_type
+            ).where(
+                self.CacheEntry.output_fingerprint_key == output_fingerprint_key,
+                self.CacheEntry.result_location != sa.null(),
+            )
+
+            match = session.execute(query).first()
+
+            if match:
+                return Path(match.result_location), match.result_type
+
+            return None, None
+
+    @staticmethod
+    def _load_from_cache(
+        path_to_cached: Path, result_type: Text
+    ) -> Tuple[Optional[Path], Optional[Type[Cacheable]]]:
+        try:
+            module = rasa.shared.utils.common.class_from_module_path(result_type)
+            assert isinstance(module, Cacheable)
+
+            return path_to_cached, module
```

Comment:
Why don't we call `module.from_cache(path_to_cached)` or similar here?

Reply:
In the sequence diagram this is done by the caller. This has the advantage that we don't have to load everything into memory right away but only when we need it.

Reply:
I.e. when we non-eagerly initialise the CachedGraphComponent?

Reply:
yep